autonat/autonat-v2.md

```diff
@@ -0,0 +1,255 @@
+
+`E_DIAL_REFUSED`: The peer didn't dial the address because of address
+based restrictions like address based rate limit, the address being a private IP
+address, or a relay address.
+
+For the address that the peer decided to dial:
+
+`E_DIAL_ERROR`: The peer was unable to connect to the address
+
+`E_CONN_UPGRADE_FAILED`: The peer was able to connect to the other address, but
+failed to complete the connection upgrade step. 
+
+`E_ATTEMPT_ERROR`: The peer was able to establish an upgraded connection but
+some error occured when sending the nonce on the `/libp2p/autonat/2/attempt`
+stream.
+
+`OK`: The peer successfully dialed the selected address.
+
+The peer MUST NOT send dial statuses for addresses after the one it selected to
+dial.
+
+
+### Requirements for ResponseStatus
+
+The `ResponseStatus` sent by the peer in the `DialResponse` message MUST be set
+according to the following requirements
+
+`OK`: the peer completed the request successfully. A request is considered
+completed successfully when the peer either completes a dial(successfully or
+unsuccessfully) or rejects all addresses in the request as undialable. 
+
+`E_BAD_REQUEST`: the peer was unable to decode the request.
+
+`E_REQUEST_REFUSED`: the peer didn't attempt to serve the request because of
+rate limiting, resource limit reached or blacklisting.
+
+`E_INTERNAL_ERROR`: error not classified within the above error codes occured on
+peer that prevented it from completing the request.
+
+Implementations MUST discard responses with status codes they do not understand. 
+
+### Amplification Attack Prevention
+
+![Interaction](autonat-v2-amplification-attack-prevention.svg)
+
+
+When a client asks a server to dial an address that is not the client's observed
+IP address, the server asks the client to send him some non trivial amount of
+bytes as a cost to dial a different IP address. To make amplification attacks
+unattractive, the number of bytes is decided such that it's sufficiently larger
+than a new connection handshake cost.
+
+On receiving a `DialRequest`, the server selects the first address it is capable
+of dialing. If this selected address has a IP different from the client's
+observed IP, the server sends a `DialDataRequest` message with the selected
+address's index(0 based) and `numBytes` set to a sufficiently large value on the
+`/libp2p/autonat/2/dial` stream
+
+Upon receiving a `DialDataRequest` message, the client decides whether to accept
+or reject the cost of dial. If the client rejects the cost, the client resets
+the stream and the `DialRequest` is considered aborted. If the client accepts
+the cost, the client starts transferring `numBytes` bytes to the server. The
+server on receiving `numBytes` bytes proceeds to dial the candidate address. 
+
+## Implementation Suggestions
+
+For any given address, implementations SHOULD do the following
+- periodically recheck reachability status
+- query multiple peers to determine reachability
```

Comment:
```suggestion
- Clients SHOULD NOT reuse their listening port when making a `DialRequest`. This avoids testing for reachability on an existing 4-tuple connection path. If both the client and server are reusing ports for dialing and listening, you'll end up using an existing hole-punched path rather than testing for reachability. 
```

Or something along those lines. The server could also control this by not reusing their listening port, but it's better for the client to do this since they can make sure they are doing the correct thing. And it avoids having the server allocate a new port.

Reply:
The server being the node that wants its address to be probed for reachability right?

How would the server allocating a new port even work? If the server is behind a router that only forwards a particular port, then the AutoNAT probe would fail despite the original address actually being reachable.

Reply:
No. The server is the node handling the dial request. The client is the initiator of the dial request. Apologies for the confusion here, I have reworded to consistently use client and server everywhere. 

>If the server is behind a router that only forwards a particular port, then the AutoNAT probe would fail despite the original address actually being reachable.

Assuming by server you mean the client that initiates the request for checking reachability.
 
This case will work fine. Say the node has setup port forwarding to forward router port 8080 to its listen port 5000. The node will not use this port to dial out, it will dial out of one of the higher ports, say 45000. The server which will handle the dial request will see the request to probe reachability for Client-IP:8080. It'll dial Client-IP:8080 and this will be forwarded correctly if it's reachable.

Reply:
@MarcoPolo I added some text regarding this. Is this fine?

Reply:
My original resolution didn't work. See [here](https://github.com/libp2p/specs/issues/562#issuecomment-1660451363) for details. Have changed this to a requirement on the server side. 

> Not sure how it works in other libp2ps, but in nim we connect to random nodes, not knowing what their capabilities are, and then we might ask them to dial-back if we find that they support autonat. We currently don't know at time of dialing if we are going to use them as AutoNat server.